src/render/src/encoder.rs

```diff
@@ -150,356 +172,373 @@ impl<B: Backend> AccessInfo<B> {
-
-        self.command_buffer.copy_buffer(
-            self.handles.ref_buffer(src.raw()).clone(),
-            self.handles.ref_buffer(dst.raw()).clone(),
-            src_offset_bytes, dst_offset_bytes, size_bytes);
-        Ok(())
     }
 
-    /// Copy part of a buffer to a texture
-    pub fn copy_buffer_to_texture_raw(
-        &mut self, src: &handle::RawBuffer<B>, src_offset_bytes: usize,
-        dst: &handle::RawTexture<B>, face: Option<texture::CubeFace>, info: texture::RawImageInfo)
-        -> CopyBufferTextureResult
-    {
-        if !src.get_info().bind.contains(memory::TRANSFER_SRC) {
-            return Err(CopyError::NoSrcBindFlag);
-        }
-        if !dst.get_info().bind.contains(memory::TRANSFER_DST) {
-            return Err(CopyError::NoDstBindFlag);
+    fn transition_image<'b>(
+        image: &'b handle::raw::Image<B>,
+        level: image::Level,
+        layer: image::Layer,
+        current: &mut core::image::State,
+        next: core::image::State,
+    ) -> Option<core::memory::Barrier<'b, B>> {
+        let state = mem::replace(current, next);
+        if state != next {
+            Some(core::memory::Barrier::Image {
+                states: state..next,
+                target: image.resource(),
+                range: (level..(level+1), layer..(layer+1)),
+            })
+        } else {
+            None
         }
+    }
 
-        let size_bytes = info.get_byte_count();
-        let src_copy_end = src_offset_bytes + size_bytes;
-        if src_copy_end > src.get_info().size {
-            return Err(CopyError::OutOfSrcBounds {
-                size: src.get_info().size,
-                copy_end: src_copy_end,
-            });
+    pub fn require_state(
+        &mut self,
+        stage: core::pso::PipelineStage,
+        buffer_states: &[(&handle::raw::Buffer<B>, core::buffer::State)],
+        image_states: &[(&handle::raw::Image<B>, image::Subresource, core::image::State)],
+    ) {
+        let mut barriers = Vec::new();
+        for &(buffer, state) in buffer_states {
+            self.require_buffer_state(buffer, state)
+                .map(|b| barriers.push(b));
         }
-
-        let dim = dst.get_info().kind.get_dimensions();
-        if !info.is_inside(dim) {
-            let (w, h, d, _) = dim;
-            return Err(CopyError::OutOfDstBounds {
-                size: [w, h, d],
-                copy_end: [info.xoffset + info.width,
-                           info.yoffset + info.height,
-                           info.zoffset + info.depth]
-            });
+        for &(image, (level, layer), state) in image_states {
+            self.require_image_state(image, level, layer, state)
+                .map(|b| barriers.push(b));
+        }
+        let current_stage = mem::replace(&mut self.pipeline_stage, stage);
+        if (current_stage != stage) || !barriers.is_empty() {
+            println!("barrier: {:?}, {:?}", current_stage..stage, &barriers[..]);
```

Comment:
please change to `debug!`

Reply:
So you will want to keep it ?

Reply:
removing works too, I don't mind either way